boofuzz/fuzzable.py

```diff
@@ -0,0 +1,135 @@
+from builtins import object
+from future.utils import listitems, with_metaclass
+
+from .mutation_context import MutationContext
+
+
+class DocStringInheritor(type):
+    """
+    A variation on
+    http://groups.google.com/group/comp.lang.python/msg/26f7b4fcb4d66c95
+    by Paul McGuire
+    """
+
+    def __new__(meta, name, bases, clsdict):
+        if not ("__doc__" in clsdict and clsdict["__doc__"]):
+            for mro_cls in (mro_cls for base in bases for mro_cls in base.mro()):
+                doc = mro_cls.__doc__
+                if doc:
+                    clsdict["__doc__"] = doc
+                    break
+        for attr, attribute in listitems(clsdict):
+            if not attribute.__doc__:
+                for mro_cls in (mro_cls for base in bases for mro_cls in base.mro() if hasattr(mro_cls, attr)):
+                    doc = getattr(getattr(mro_cls, attr), "__doc__")
+                    if doc:
+                        if isinstance(attribute, property):
+                            clsdict[attr] = property(attribute.fget, attribute.fset, attribute.fdel, doc)
+                        else:
+                            attribute.__doc__ = doc
+                        break
+        return type.__new__(meta, name, bases, clsdict)
+
+
+# DocStringInheritor is the metaclass in python 2 and 3
+class Fuzzable(with_metaclass(DocStringInheritor, object)):
+    """Base class for fuzzable message element types.
+
+    A typical Fuzzable type will implement mutations() (a generator) or encode() or both.
+
+    num_mutations() exists as an optimization -- if exhausting mutations is inefficient, one may provide num_mutations()
+    to compute the number of mutations more efficiently.
+
+    It's OK if your muations() function yields an indeterminate number of mutations -- it will just mess up the stats
+    while the fuzzer is running.
+    """
+
+    def __init__(self):
+        self._request = None
+        self._context_path = None
+
+    def mutations(self):
+        """Generator to yield mutation values for this element.
+
+        Values are either plain values or callable functions that take a "default value" and mutate it. Functions are
+        used when the default or "normal" value influences the fuzzed value. Functions are used because the "normal"
+        value is sometimes dynamic and not known at the time of generation.
+
+        Each mutation should be a pre-rendered value. That is, it must be suitable to pass to encode().
+
+        Default: Empty iterator.
+        """
+        return
+        yield
```

Comment:
redudant statement `return`? Or a magic trick?

Reply:
Empty generator trick. See https://stackoverflow.com/questions/13243766/python-empty-generator-function

Now that I read that answer again, I think `return iter(())` is more intuitive.

Reply:
Wow, thanks, got a new trick :)